File: jobmaster/summary.py
from __future__ import annotations

import re
from typing import Any


def _clean(value: str) -> str:
    return re.sub(r"\s+", " ", value).strip(" ,.;")


def _split_multi(value: str) -> list[str]:
    if not value:
        return []
    chunks = re.split(r"[|,/;]\s*|\s{2,}", value)
    return [_clean(chunk) for chunk in chunks if _clean(chunk)]


def _split_bio_parts(value: str) -> list[str]:
    return [_clean(chunk) for chunk in value.split("|") if _clean(chunk)]


def _unique(values: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for value in values:
        key = value.lower()
        if key in seen:
            continue
        seen.add(key)
        result.append(value)
    return result


def _natural_join(values: list[str]) -> str:
    values = [value for value in values if value]
    if not values:
        return ""
    if len(values) == 1:
        return values[0]
    if len(values) == 2:
        return f"{values[0]} and {values[1]}"
    return f"{', '.join(values[:-1])}, and {values[-1]}"


def _bio_identity(github_bio: str) -> str:
    first = _split_bio_parts(github_bio)[:1]
    if not first:
        return ""
    lead = first[0]
    if "@" in lead:
        role, org = [piece.strip() for piece in lead.split("@", 1)]
        return f"{org} {role.lower()} student"
    return lead


def _bio_focus_areas(github_bio: str) -> tuple[list[str], str]:
    parts = _split_bio_parts(github_bio)
    focus: list[str] = []
    seeking = ""
    for part in parts[1:]:
        if part.lower().startswith("seeking "):
            seeking = part
        else:
            focus.extend(_split_multi(part))
    return focus, seeking


def _repo_languages(candidate_sources: dict[str, Any]) -> list[str]:
    repositories = candidate_sources.get("github", {}).get("repositories", [])
    langs = [
        str(repo.get("language", "")).strip()
        for repo in repositories
        if str(repo.get("language", "")).strip() and str(repo.get("language", "")).strip().lower() != "other"
    ]
    return _unique(langs)


def _sentence_open(value: str) -> str:
    if not value:
        return ""
    return value[0].upper() + value[1:]


def generate_professional_summary(profile: dict[str, Any], candidate_sources: dict[str, Any]) -> str:
    github = candidate_sources.get("github", {})
    preferences = candidate_sources.get("preferences", {})
    education = candidate_sources.get("education", {})
    github_bio = str(github.get("bio", "")).strip()

    identity = ""
    if education.get("school") and education.get("degree"):
        identity = f"{education['school']} {str(education['degree']).lower()} student"
    elif education.get("school"):
        identity = f"Student at {education['school']}"
    else:
        identity = _bio_identity(github_bio)

    bio_focus, seeking = _bio_focus_areas(github_bio)
    skills = _unique(
        bio_focus
        + _split_multi(str(profile.get("top_skills", "")))
        + _repo_languages(candidate_sources)
        + [str(candidate_sources.get("preferences", {}).get("industries", "")).strip()]
    )
    skills = [skill for skill in skills if skill and not skill.lower().startswith("seeking ")]

    project_highlights = candidate_sources.get("project_highlights", [])
    repo_count = int(github.get("repository_count", 0) or 0)
    repos = github.get("repositories", [])
    repo_descriptions = [str(repo.get("description", "")).strip() for repo in repos if str(repo.get("description", "")).strip()]

    sentence_1 = ""
    filtered_skills = [skill for skill in skills if "internship" not in skill.lower()]
    if identity and filtered_skills:
        sentence_1 = f"{_sentence_open(identity)} with hands-on experience in {_natural_join(filtered_skills[:4])}."
    elif identity:
        sentence_1 = f"{_sentence_open(identity)}."
    else:
        sentence_1 = "Hands-on candidate with practical technical project experience."

    sentence_2 = ""
    if project_highlights:
        sentence_2 = f"Recent work includes {_natural_join([str(item) for item in project_highlights[:2]])}."
    elif repo_descriptions:
        useful_descriptions = [item for item in repo_descriptions if not item.lower().startswith("forked coursework")]
        sentence_2 = (
            f"Recent work includes {_natural_join(useful_descriptions[:2])}, alongside {repo_count} public GitHub repositories"
            if repo_count
            else f"Recent work includes {_natural_join(useful_descriptions[:2])}"
        )
        sentence_2 = sentence_2.rstrip(".") + "."
    elif repo_count:
        sentence_2 = f"Maintains {repo_count} public GitHub repositories that showcase applied project work and ongoing technical learning."

    target_roles = str(preferences.get("target_roles", "")).strip()
    sentence_3 = ""
    if target_roles:
        sentence_3 = f"Interested in {target_roles}."
    elif seeking:
        sentence_3 = seeking.rstrip(".") + "."

    summary = " ".join(part for part in [sentence_1, sentence_2, sentence_3] if part).strip()
    return re.sub(r"\s+", " ", summary)

File: extension/src/shared/summary.js
function clean(value) {
  return String(value ?? "").replace(/\s+/g, " ").trim().replace(/^[,.;\s]+|[,.;\s]+$/g, "");
}

function splitMulti(value) {
  if (!value) {
    return [];
  }
  return String(value)
    .split(/[|,/;]\s*|\s{2,}/)
    .map(clean)
    .filter(Boolean);
}

function splitBioParts(value) {
  return String(value ?? "")
    .split("|")
    .map(clean)
    .filter(Boolean);
}

function unique(values) {
  const seen = new Set();
  const result = [];
  for (const value of values) {
    const key = String(value).toLowerCase();
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    result.push(value);
  }
  return result;
}

function naturalJoin(values) {
  const cleanValues = values.filter(Boolean);
  if (!cleanValues.length) {
    return "";
  }
  if (cleanValues.length === 1) {
    return cleanValues[0];
  }
  if (cleanValues.length === 2) {
    return `${cleanValues[0]} and ${cleanValues[1]}`;
  }
  return `${cleanValues.slice(0, -1).join(", ")}, and ${cleanValues.at(-1)}`;
}

function bioIdentity(githubBio) {
  const lead = splitBioParts(githubBio)[0];
  if (!lead) {
    return "";
  }
  if (lead.includes("@")) {
    const [role, org] = lead.split("@").map((part) => part.trim());
    return `${org} ${role.toLowerCase()} student`;
  }
  return lead;
}

function bioFocusAreas(githubBio) {
  const parts = splitBioParts(githubBio);
  const focus = [];
  let seeking = "";
  for (const part of parts.slice(1)) {
    if (part.toLowerCase().startsWith("seeking ")) {
      seeking = part;
    } else {
      focus.push(...splitMulti(part));
    }
  }
  return { focus, seeking };
}

function repoLanguages(candidateSources) {
  const repos = candidateSources?.github?.repositories ?? [];
  return unique(
    repos
      .map((repo) => clean(repo.language))
      .filter((language) => language && language.toLowerCase() !== "other")
  );
}

function sentenceOpen(value) {
  if (!value) {
    return "";
  }
  return value[0].toUpperCase() + value.slice(1);
}

export function generateProfessionalSummary(profile = {}, candidateSources = {}) {
  const github = candidateSources.github ?? {};
  const preferences = candidateSources.preferences ?? {};
  const education = candidateSources.education ?? {};
  const githubBio = clean(github.bio);

  let identity = "";
  if (education.school && education.degree) {
    identity = `${education.school} ${String(education.degree).toLowerCase()} student`;
  } else if (education.school) {
    identity = `Student at ${education.school}`;
  } else {
    identity = bioIdentity(githubBio);
  }

  const { focus, seeking } = bioFocusAreas(githubBio);
  const skills = unique(
    [
      ...focus,
      ...splitMulti(profile.top_skills),
      ...repoLanguages(candidateSources),
      clean(preferences.industries)
    ].filter((value) => value && !value.toLowerCase().startsWith("seeking "))
  );

  const filteredSkills = skills.filter((skill) => !skill.toLowerCase().includes("internship"));
  const projectHighlights = candidateSources.project_highlights ?? [];
  const repoDescriptions = (github.repositories ?? [])
    .map((repo) => clean(repo.description))
    .filter((value) => value && !value.toLowerCase().startsWith("forked coursework"));
  const repoCount = Number(github.repository_count ?? 0);

  let sentence1 = "";
  if (identity && filteredSkills.length) {
    sentence1 = `${sentenceOpen(identity)} with hands-on experience in ${naturalJoin(filteredSkills.slice(0, 4))}.`;
  } else if (identity) {
    sentence1 = `${sentenceOpen(identity)}.`;
  } else {
    sentence1 = "Hands-on candidate with practical technical project experience.";
  }

  let sentence2 = "";
  if (projectHighlights.length) {
    sentence2 = `Recent work includes ${naturalJoin(projectHighlights.slice(0, 2).map(String))}.`;
  } else if (repoDescriptions.length) {
    sentence2 = repoCount
      ? `Recent work includes ${naturalJoin(repoDescriptions.slice(0, 2))}, alongside ${repoCount} public GitHub repositories.`
      : `Recent work includes ${naturalJoin(repoDescriptions.slice(0, 2))}.`;
  } else if (repoCount) {
    sentence2 = `Maintains ${repoCount} public GitHub repositories that showcase applied project work and ongoing technical learning.`;
  }

  const targetRoles = clean(preferences.target_roles);
  let sentence3 = "";
  if (targetRoles) {
    sentence3 = `Interested in ${targetRoles}.`;
  } else if (seeking) {
    sentence3 = `${seeking.replace(/[.]+$/, "")}.`;
  }

  return [sentence1, sentence2, sentence3].filter(Boolean).join(" ").replace(/\s+/g, " ").trim();
}

